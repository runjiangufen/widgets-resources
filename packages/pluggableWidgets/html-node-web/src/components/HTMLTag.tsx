import { createElement, HTMLAttributes, ReactElement, ReactNode } from "react";

interface HTMLTagProps {
    tagName: keyof JSX.IntrinsicElements;
    unsafeHTML?: string;
    children: ReactNode;
    attributes: HTMLAttributes<Element> & { [dataAttribute: `data-${string}`]: string };
}

export function HTMLTag(props: HTMLTagProps): ReactElement {
    const Tag = props.tagName;
    const { unsafeHTML } = props;
    if (unsafeHTML !== undefined) {
        return <Tag {...props.attributes} dangerouslySetInnerHTML={{ __html: unsafeHTML }} />;
    }

    return <Tag {...props.attributes}>{props.children}</Tag>;
}
