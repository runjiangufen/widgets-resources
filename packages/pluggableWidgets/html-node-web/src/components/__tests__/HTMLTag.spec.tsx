import { render, fireEvent } from "@testing-library/react";
import { createElement } from "react";

import { HTMLTag } from "../HTMLTag";

describe("HTMLTag", () => {
    it("renders correctly with innerHTML", () => {
        const { asFragment } = render(
            <HTMLTag
                tagName="div"
                unsafeHTML="<p>Lorem ipsum</p>"
                attributes={{
                    className: "html-node-root my-class"
                }}
            >
                {undefined}
            </HTMLTag>
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("renders correctly with content", () => {
        const { asFragment } = render(
            <HTMLTag
                tagName="div"
                unsafeHTML={undefined}
                attributes={{
                    className: "html-node-root my-another-class",
                    style: { color: "red" }
                }}
            >
                <div>Lorem ipsum</div>
            </HTMLTag>
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("fires events", () => {
        const cbFn = jest.fn();
        const { getByTestId } = render(
            <HTMLTag
                tagName="div"
                unsafeHTML={undefined}
                attributes={{
                    "data-testid": "html-node",
                    onClick: _e => {
                        cbFn();
                    }
                }}
            >
                <div>Lorem ipsum</div>
            </HTMLTag>
        );

        fireEvent.click(getByTestId("html-node"));

        expect(cbFn).toHaveBeenCalled();
    });
});
