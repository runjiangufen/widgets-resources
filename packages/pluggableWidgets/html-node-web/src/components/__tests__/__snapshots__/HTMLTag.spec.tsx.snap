// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HTMLTag renders correctly with content 1`] = `
<DocumentFragment>
  <div
    class="html-node-root my-another-class"
    style="color: red;"
  >
    <div>
      Lorem ipsum
    </div>
  </div>
</DocumentFragment>
`;

exports[`HTMLTag renders correctly with innerHTML 1`] = `
<DocumentFragment>
  <div
    class="html-node-root my-class"
  >
    <p>
      Lorem ipsum
    </p>
  </div>
</DocumentFragment>
`;
